### PR TITLE
Update contrib/re2 to 2023-06-02

### PR DESCRIPTION
**Pull request by georgthegreat**
### Changelog category (leave one):
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):

* Update contrib/re2 to 2023-06-02

### CODE REVIEW

**Comment by georgthegreat** [comment]
Let's run the CI and see if this will work.
(this worked for us internally, but our setup might be different)

**Comment by georgthegreat** [comment]
I believe this should work as ClickHouse should be using this branch of absl:

```
#ifdef ABSL_USES_STD_STRING_VIEW

#include <string_view>  // IWYU pragma: export

namespace absl {
ABSL_NAMESPACE_BEGIN
using string_view = std::string_view;
ABSL_NAMESPACE_END
}  // namespace absl

#else  // ABSL_USES_STD_STRING_VIEW
```

**Comment by robot-ch-test-poll4** [comment]
This is an automated comment for commit 9a5c5c7e931c41f406e316cb20cc7659c40fccc5 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50949/9a5c5c7e931c41f406e316cb20cc7659c40fccc5/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by georgthegreat** [comment]
@nickitat, CMake build fails with

```
2023-06-13 21:49:11 CMake Error at contrib/re2-cmake/CMakeLists.txt:38 (add_library):
2023-06-13 21:49:11   No SOURCES given to target: re2
2023-06-13 21:49:11 
2023-06-13 21:49:11 
2023-06-13 21:49:11 CMake Error at contrib/re2-cmake/CMakeLists.txt:46 (add_library):
2023-06-13 21:49:11   No SOURCES given to target: re2_st
```

Should I change something except for simply updating git submodule?

[submodule_log](https://s3.amazonaws.com/clickhouse-test-reports/50949/9a5c5c7e931c41f406e316cb20cc7659c40fccc5/fast_test/submodule_log.txt) and [clone_log](https://s3.amazonaws.com/clickhouse-test-reports/50949/9a5c5c7e931c41f406e316cb20cc7659c40fccc5/fast_test/clone_log.txt) look good.

File: contrib/re2-cmake/CMakeLists.txt [hunk 1]
```diff
@@ -59,17 +65,6 @@ foreach (FILENAME filtered_re2.h re2.h set.h stringpiece.h)
     add_dependencies (re2_st transform_${FILENAME})
 endforeach ()
 
-file (MAKE_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}/util)
```
Comment (nickitat) [inline contrib/re2-cmake/CMakeLists.txt:63]: mutex [was deleted](https://github.com/google/re2/commit/49d776b9d29d79b6e2876d5f091d2207d8123dfa)

File: contrib/abseil-cpp-cmake/CMakeLists.txt [hunk 2]
```diff
@@ -17,3 +17,17 @@ get_target_property(FLAT_HASH_SET_INCLUDE_DIR absl::flat_hash_set INTERFACE_INCL
 target_include_directories (_abseil_swiss_tables SYSTEM BEFORE INTERFACE ${FLAT_HASH_SET_INCLUDE_DIR})
 
 add_library(ch_contrib::abseil_swiss_tables ALIAS _abseil_swiss_tables)
+
+set(ABSL_FORMAT_SRC
+    ${ABSL_ROOT_DIR}/absl/strings/internal/str_format/arg.cc
```
Comment (rschu1ze) [inline contrib/abseil-cpp-cmake/CMakeLists.txt:22]: This lib looks roughly modeled after the `str_format_internal` lib in contrib/abseil-cpp/absl/strings/CMakeLists.txt. Maybe add a comment a la `/// modeled after str_format_internal in native build system`

File: contrib/re2-cmake/CMakeLists.txt [hunk 3]
```diff
@@ -48,6 +50,7 @@ target_compile_definitions (re2_st PRIVATE NDEBUG NO_THREADS re2=re2_st)
 target_include_directories (re2_st PRIVATE .)
 target_include_directories (re2_st SYSTEM PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
 target_include_directories (re2_st SYSTEM BEFORE PUBLIC ${SRC_DIR})
+target_link_libraries (re2_st ch_contrib::abseil_str_format)
```
Comment (rschu1ze) [inline contrib/re2-cmake/CMakeLists.txt:53]: I know it has been like that but the transformation code (l. 55ff) looks very weird and fragile. The reader wonders:
- why transform only the four files given in l. 56? What about the other files?
- is the original build system doing the same stuff or can the library variant (re2 vs. re2_st) be selected via a simple preprocessor define?

File: src/Common/parseGlobs.cpp [hunk 4]
```diff
@@ -40,7 +40,7 @@ std::string makeRegexpPatternFromGlobs(const std::string & initial_str_with_glob
     size_t current_index = 0;
     while (RE2::FindAndConsume(&input, enum_or_range, &matched))
     {
-        std::string buffer = matched.ToString();
+        std::string buffer{matched};
```
[the remark this answers is not in the document]
Comment (georgthegreat) [reply src/Common/parseGlobs.cpp:43]: This PR turned out to be more about version switch / compatibility.
Further updates / optimizations might become possible.

I suggest reviewing them later on.

[the remark this answers is not in the document]
Comment (georgthegreat) [reply src/Common/parseGlobs.cpp:43]: Plus I don't quite undestand how can I pull @nickitat patches into my local branch with git cli.

**Comment by georgthegreat** [comment]
@nickitat, the CI looks good.

Let's merge it. I will review the code in search of possible improvements after the the merge (though I doubt there will be any big one).

**Comment by alexey-milovidov** [comment]
All good, but we have to fix one integration test...

**Comment by georgthegreat** [comment]
@alexey-milovidov, all green!

File: src/Common/parseGlobs.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rschu1ze) [reply src/Common/parseGlobs.cpp:43]: Made a quick follow-up: https://github.com/ClickHouse/ClickHouse/pull/51655

**Pull request merged**